File: code/ts/test/parity.test.ts
/// Parity tests — validates TypeScript SDK against shared JSON fixtures.
///
/// Reads fixtures from code/tests/fixtures/ and verifies that
/// the TS implementation produces identical results to Dart and Python.
import { describe, it, expect } from 'vitest';
import { readFileSync } from 'node:fs';
import { resolve } from 'node:path';
import { success, failure, isSuccess, isFailure } from '../src/core/result';
import { ServiceFailure } from '../src/core/service_failure';
import { ServiceClientConfig, ServiceRequest } from '../src/core/service_client';
import { HttpClientException } from '../src/http/http_client_exception';

type FixtureCase = Record<string, unknown>;
interface Fixture {
  cases: FixtureCase[];
}

function loadFixture(name: string): Fixture {
  const path = resolve(__dirname, '../../tests/fixtures', `${name}.json`);
  return JSON.parse(readFileSync(path, 'utf-8')) as Fixture;
}

describe('Parity: Result', () => {
  const fixture = loadFixture('result');

  for (const c of fixture.cases) {
    const id = c['id'] as string;
    const input = c['input'] as Record<string, unknown>;
    const expected = c['expect'] as Record<string, unknown>;

    it(id, () => {
      if (input['type'] === 'success') {
        const result = success(input['value']);
        expect(isSuccess(result)).toBe(expected['is_success']);
        expect(isFailure(result)).toBe(expected['is_failure']);
        if (isSuccess(result)) {
          expect(result.value).toEqual(expected['value']);
        }
      } else {
        if (typeof input['error'] === 'object' && input['error'] !== null) {
          const errorMap = input['error'] as Record<string, unknown>;
          const sf = new ServiceFailure({
            statusCode: errorMap['status_code'] as number,
            message: errorMap['message'] as string,
          });
          const result = failure(sf);
          expect(isFailure(result)).toBe(true);
          if (isFailure(result)) {
            const err = result.error as ServiceFailure;
            expect(err.statusCode).toBe(expected['error_status_code']);
            expect(err.message).toBe(expected['error_message']);
          }
        } else {
          const result = failure(input['error']);
          expect(isFailure(result)).toBe(true);
          if (isFailure(result)) {
            expect(result.error).toEqual(expected['error']);
          }
        }
      }
    });
  }
});

describe('Parity: ServiceClientConfig', () => {
  const fixture = loadFixture('service_client_config');

  for (const c of fixture.cases) {
    const id = c['id'] as string;
    const input = c['input'] as Record<string, unknown>;
    const expected = c['expect'] as Record<string, unknown>;

    it(id, () => {
      const config = new ServiceClientConfig({
        baseUrl: input['base_url'] as string,
      });

      if ('base_url_ends_with_slash' in expected) {
        expect(config.baseUrl.pathname.endsWith('/')).toBe(
          expected['base_url_ends_with_slash'],
        );
      }
      if ('timeout_seconds' in expected) {
        expect(config.timeout / 1000).toBe(expected['timeout_seconds']);
      }
      if ('default_headers' in expected) {
        expect(config.defaultHeaders).toEqual(expected['default_headers']);
      }
    });
  }
});

describe('Parity: ServiceRequest', () => {
  const fixture = loadFixture('service_request');

  for (const c of fixture.cases) {
    const id = c['id'] as string;
    const input = c['input'] as Record<string, unknown>;
    const expected = c['expect'] as Record<string, unknown>;

    it(id, () => {
      const request = ServiceRequest.http({
        method: input['method'] as string,
        endpoint: input['endpoint'] as string,
        headers: input['headers'] as Record<string, string> | undefined,
        body: input['body'] as unknown,
        errorMessage: (input['error_message'] as string) ?? undefined,
      });

      if ('protocol' in expected) {
        expect(request.protocol).toBe(expected['protocol']);
      }
      if ('error_message' in expected) {
        expect(request.errorMessage).toBe(expected['error_message']);
      }
      if ('method' in expected) {
        expect(request.method).toBe(expected['method']);
      }
      if ('endpoint' in expected) {
        expect(request.endpoint).toBe(expected['endpoint']);
      }
      if ('headers' in expected) {
        expect(request.headers).toEqual(expected['headers']);
      }
      if ('body' in expected) {
        expect(request.body).toEqual(expected['body']);
      }
    });
  }
});

describe('Parity: Errors', () => {
  const fixture = loadFixture('errors');

  for (const c of fixture.cases) {
    const id = c['id'] as string;
    const input = c['input'] as Record<string, unknown>;
    const expected = c['expect'] as Record<string, unknown>;

    it(id, () => {
      if ((id as string).startsWith('service-failure')) {
        const sf = new ServiceFailure({
          statusCode: input['status_code'] as number,
          message: input['message'] as string,
          responseBody: input['response_body'] as Record<string, unknown> | undefined,
        });

        expect(sf.statusCode).toBe(input['status_code']);
        expect(sf.message).toBe(input['message']);

        if ('response_body_is_null' in expected) {
          expect(sf.responseBody).toBeUndefined();
        }
        if ('response_body' in expected) {
          expect(sf.responseBody).toEqual(expected['response_body']);
        }
        if ('to_string_contains' in expected) {
          const str = sf.toString();
          for (const s of expected['to_string_contains'] as string[]) {
            expect(str).toContain(s);
          }
        }
      } else {
        const exc = new HttpClientException({
          statusCode: input['status_code'] as number,
          message: input['message'] as string,
          response: input['response'] as Record<string, unknown> | undefined,
        });

        expect(exc.statusCode).toBe(input['status_code']);
        expect(exc.message).toBe(input['message']);

        if ('response_is_null' in expected) {
          expect(exc.response).toBeUndefined();
        }
        if ('response' in expected) {
          expect(exc.response).toEqual(expected['response']);
        }
        if ('to_string_contains' in expected) {
          const str = exc.toString();
          for (const s of expected['to_string_contains'] as string[]) {
            expect(str).toContain(s);
          }
        }
      }
    });
  }
});

File: code/dart/test/parity_test.dart
/// Parity tests — validates Dart SDK against shared JSON fixtures.
///
/// These tests read fixtures from code/tests/fixtures/ and verify that
/// the Dart implementation produces identical results to TS and Python.
import 'dart:convert';
import 'dart:io';

import 'package:test/test.dart';
import 'package:service_client/service_client.dart';

Map<String, dynamic> _loadFixture(String name) {
  final path = '../tests/fixtures/$name.json';
  final file = File(path);
  return jsonDecode(file.readAsStringSync()) as Map<String, dynamic>;
}

void main() {
  group('Parity: Result', () {
    final fixture = _loadFixture('result');
    final cases = fixture['cases'] as List;

    for (final c in cases) {
      final id = c['id'] as String;
      final input = c['input'] as Map<String, dynamic>;
      final expect_ = c['expect'] as Map<String, dynamic>;

      test(id, () {
        if (input['type'] == 'success') {
          final result = Result<dynamic, dynamic>.success(input['value']);
          expect(result, isA<Success>());
          expect((result as Success).value, equals(expect_['value']));
          if (expect_.containsKey('is_success')) {
            expect(result is Success, equals(expect_['is_success']));
          }
        } else {
          if (input['error'] is Map) {
            final errorMap = input['error'] as Map<String, dynamic>;
            final sf = ServiceFailure(
              statusCode: errorMap['status_code'] as int,
              message: errorMap['message'] as String,
            );
            final result = Result<dynamic, ServiceFailure>.failure(sf);
            expect(result, isA<Failure>());
            final err = (result as Failure).error;
            expect(err.statusCode, equals(expect_['error_status_code']));
            expect(err.message, equals(expect_['error_message']));
          } else {
            final result =
                Result<dynamic, dynamic>.failure(input['error']);
            expect(result, isA<Failure>());
            expect((result as Failure).error, equals(expect_['error']));
          }
        }
      });
    }
  });

  group('Parity: ServiceClientConfig', () {
    final fixture = _loadFixture('service_client_config');
    final cases = fixture['cases'] as List;

    for (final c in cases) {
      final id = c['id'] as String;
      final input = c['input'] as Map<String, dynamic>;
      final expect_ = c['expect'] as Map<String, dynamic>;

      test(id, () {
        final config = ServiceClientConfig(
          baseUrl: Uri.parse(input['base_url'] as String),
        );

        if (expect_.containsKey('base_url_ends_with_slash')) {
          final path = config.baseUrl.path;
          expect(
            path.isEmpty || path.endsWith('/'),
            equals(expect_['base_url_ends_with_slash']),
          );
        }
        if (expect_.containsKey('timeout_seconds')) {
          expect(
            config.timeout.inSeconds,
            equals(expect_['timeout_seconds']),
          );
        }
        if (expect_.containsKey('default_headers')) {
          expect(config.defaultHeaders, equals(expect_['default_headers']));
        }
      });
    }
  });

  group('Parity: ServiceRequest', () {
    final fixture = _loadFixture('service_request');
    final cases = fixture['cases'] as List;

    for (final c in cases) {
      final id = c['id'] as String;
      final input = c['input'] as Map<String, dynamic>;
      final expect_ = c['expect'] as Map<String, dynamic>;

      test(id, () {
        final request = ServiceRequest.http(
          method: input['method'] as String,
          endpoint: input['endpoint'] as String,
          headers: input.containsKey('headers')
              ? Map<String, String>.from(input['headers'] as Map)
              : null,
          body: input['body'],
          errorMessage: input.containsKey('error_message')
              ? input['error_message'] as String
              : 'Error in HTTP request',
        );

        if (expect_.containsKey('protocol')) {
          expect(request.protocol, equals(ServiceProtocol.http));
        }
        if (expect_.containsKey('error_message')) {
          expect(request.errorMessage, equals(expect_['error_message']));
        }
        if (expect_.containsKey('method')) {
          expect(request.method, equals(expect_['method']));
        }
        if (expect_.containsKey('endpoint')) {
          expect(request.endpoint, equals(expect_['endpoint']));
        }
        if (expect_.containsKey('headers')) {
          expect(request.headers, equals(expect_['headers']));
        }
        if (expect_.containsKey('body')) {
          expect(request.body, equals(expect_['body']));
        }
      });
    }
  });

  group('Parity: Errors', () {
    final fixture = _loadFixture('errors');
    final cases = fixture['cases'] as List;

    for (final c in cases) {
      final id = c['id'] as String;
      final input = c['input'] as Map<String, dynamic>;
      final expect_ = c['expect'] as Map<String, dynamic>;

      test(id, () {
        if (id.startsWith('service-failure')) {
          final sf = ServiceFailure(
            statusCode: input['status_code'] as int,
            message: input['message'] as String,
            responseBody: input.containsKey('response_body')
                ? Map<String, dynamic>.from(input['response_body'] as Map)
                : null,
          );

          expect(sf.statusCode, equals(input['status_code']));
          expect(sf.message, equals(input['message']));

          if (expect_.containsKey('response_body_is_null')) {
            expect(sf.responseBody, isNull);
          }
          if (expect_.containsKey('response_body')) {
            expect(sf.responseBody, equals(expect_['response_body']));
          }
          if (expect_.containsKey('to_string_contains')) {
            final str = sf.toString();
            for (final s in expect_['to_string_contains']) {
              expect(str, contains(s));
            }
          }
        } else {
          final exc = HttpClientException(
            statusCode: input['status_code'] as int,
            message: input['message'] as String,
            response: input.containsKey('response')
                ? Map<String, dynamic>.from(input['response'] as Map)
                : null,
          );

          expect(exc.statusCode, equals(input['status_code']));
          expect(exc.message, equals(input['message']));

          if (expect_.containsKey('response_is_null')) {
            expect(exc.response, isNull);
          }
          if (expect_.containsKey('response')) {
            expect(exc.response, equals(expect_['response']));
          }
          if (expect_.containsKey('to_string_contains')) {
            final str = exc.toString();
            for (final s in expect_['to_string_contains']) {
              expect(str, contains(s));
            }
          }
        }
      });
    }
  });
}
